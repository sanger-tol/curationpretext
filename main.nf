#!/usr/bin/env nextflow
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    sanger-tol/curationpretext
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Github : https://github.com/sanger-tol/curationpretext
    Website: https://nf-co.re/curationpretext
    Slack  : https://nfcore.slack.com/channels/curationpretext
----------------------------------------------------------------------------------------
*/

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT FUNCTIONS / MODULES / SUBWORKFLOWS / WORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

include { CURATIONPRETEXT_ALLF      } from './workflows/curationpretext_allf'
//include { CURATIONPRETEXT_MAPS      } from './workflows/curationpretext_maps'
include { PIPELINE_INITIALISATION   } from './subworkflows/local/utils_nfcore_curationpretext_pipeline'
include { PIPELINE_COMPLETION       } from './subworkflows/local/utils_nfcore_curationpretext_pipeline'
include { getGenomeAttribute        } from './subworkflows/local/utils_nfcore_curationpretext_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    NAMED WORKFLOWS FOR PIPELINE
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/


//
// WORKFLOW: Run main sanger-tol/curationpretext analysis pipeline
//
workflow SANGER_TOL_CURATIONPRETEXT {
    take:
    input_fasta
    reads
    cram
    sample
    teloseq
    aligner
    read_type
    map_order

    main:

    CURATIONPRETEXT_ALLF (
        input_fasta,
        reads,
        cram,
        sample,
        teloseq,
        aligner,
        read_type,
        map_order
    )
    // CURATIONPRETEXT_MAPS
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow {

    //
    // SUBWORKFLOW: Run initialisation tasks
    //
    PIPELINE_INITIALISATION (
        params.version,
        params.validate_params,
        params.monochrome_logs,
        args,
        params.outdir,
        []                      // We are not using the samplesheet for this pipeline
    )

    // MOVE THE CHANNEL CREATION INTO THE PIPELINE INITIALISATION

    //
    // WORFKLOW: Run main sanger-tol/curationpretext analysis pipeline
    //
    SANGER_TOL_CURATIONPRETEXT (
        params.input,
        params.reads,
        params.cram,
        params.sample,
        params.teloseq,
        params.aligner,
        params.read_type,
        params.map_order
    )

    //
    // SUBWORKFLOW: Run completion tasks
    //
    PIPELINE_COMPLETION (
        params.email,
        params.email_on_fail,
        params.plaintext_email,
        params.outdir,
        params.monochrome_logs,
        params.hook_url,
        []                      // We are not using MultiQC for this pipeline
    )
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
