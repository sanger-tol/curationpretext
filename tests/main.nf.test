nextflow_pipeline {

    name "Test Workflow main.nf"
    script "main.nf"
    tag "pipeline"
    tag "pipeline_sanger_tol"
    tag "full"

    test("Full run") {

        when {
            params {
                outdir = "${outputDir}"
                all_output = true
                skip_tracks = "NONE"
                split_telomere = true
            }
        }

        then {
            def stable_name = getAllFilesFromDir(params.outdir, relative: true, includeDir: true, ignore: ["pipeline_info/*.{html,json,txt}"])

            def pretext_files = getAllFilesFromDir(params.outdir, include: ['pretext_maps_processed/*.pretext'])

            def pretext_maps_raw = getAllFilesFromDir(params.outdir, include: ['pretext_maps_raw/*.pretext'])

            def pretext_png = getAllFilesFromDir(params.outdir, include: ['pretext_snapshot/*.png'])

            def accessories = getAllFilesFromDir(params.outdir, include: ['accessory_files/*'])

            assertAll(
                {assert workflow.success},
                {assert snapshot(
                    // Test for number of successful processes - should be 29 for a full run
                    workflow.trace.succeeded().size(),

                    removeNextflowVersion("$outputDir/pipeline_info/sanger-tol_curationpretext_software_versions.yml"),

                    // Stable name with relative path
                    stable_name,
                    stable_name.size(),

                    // Accessory files
                    accessories,
                    accessories.size(),

                    // The two pretext files
                    // Presence of files indicated presence of the raw_pretexts
                    // we expect this to be a list of two files
                    // we can't use their md5sum as they will be different everytime
                    // Then double check that there are two
                    // one is a hr and the other a normal variant
                    pretext_maps_raw.size(),
                    pretext_maps_raw.any{it.toString().contains("_hr_pi.pretext".toString())},
                    pretext_maps_raw.any{it.toString().contains("_normal_pi.pretext".toString())},

                    pretext_files.size(),
                    pretext_files.any{it.toString().contains("_hr.pretext".toString())},
                    pretext_files.any{it.toString().contains("_normal.pretext".toString())},

                    // The one png file - like the maps, it will be slightly different
                    pretext_png.size(),
                    pretext_png.any{it.toString().contains(".png".toString())}

                ).match()},
            )
        }
    }
}
