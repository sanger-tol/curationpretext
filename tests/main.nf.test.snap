{
    "Full run": {
        "content": [
            43,
            {
                "BEDTOOLS_BAMTOBED": {
                    "bedtools": "2.31.1"
                },
                "BEDTOOLS_GENOMECOV": {
                    "bedtools": "2.31.1"
                },
                "BEDTOOLS_INTERSECT": {
                    "bedtools": "2.31.1"
                },
                "BEDTOOLS_MAKEWINDOWS": {
                    "bedtools": "2.31.1"
                },
                "BEDTOOLS_MAP": {
                    "bedtools": "2.31.1"
                },
                "BEDTOOLS_MERGE_MIN": {
                    "bedtools": "2.31.1"
                },
                "BWAMEM2_INDEX": {
                    "bwamem2": "2.2.1"
                },
                "CRAM_FILTER_ALIGN_BWAMEM2_FIXMATE_SORT": {
                    "samtools": 1.17,
                    "bwa-mem2": "2.2.1",
                    "staden_io_lib": "1.14.14"
                },
                "EXTRACT_REPEAT": {
                    "perl": "(v5.26.2))",
                    "extract_repeat.pl": 1.0
                },
                "EXTRACT_TELO": {
                    "extract_telomere": 2.0,
                    "coreutils": 9.1
                },
                "FINDHALFCOVERAGE": {
                    "python": "Python 3.9.1",
                    "findHalfcoverage.py": 1.0
                },
                "FIND_TELOMERE_REGIONS": {
                    "find_telomere": null
                },
                "FIND_TELOMERE_WINDOWS": {
                    "telomere": 1.0
                },
                "GAWK_GAP_LENGTH": {
                    "gawk": "5.3.0"
                },
                "GAWK_GENERATE_GENOME_FILE": {
                    "gawk": "5.3.0"
                },
                "GAWK_REFORMAT_INTERSECT": {
                    "gawk": "5.3.0"
                },
                "GAWK_RENAME_IDS": {
                    "gawk": "5.3.0"
                },
                "GAWK_REPLACE_DOTS": {
                    "gawk": "5.3.0"
                },
                "GENERATE_CRAM_CSV": {
                    "samtools": 1.17
                },
                "GETMINMAXPUNCHES": {
                    "getminmaxpunches": 2.0,
                    "coreutils": 9.1
                },
                "GET_LARGEST_SCAFF": {
                    "get_largest_scaff": 2.0,
                    "coreutils": 9.1
                },
                "GNU_SORT": {
                    "coreutils": 9.3
                },
                "GNU_SORT_A": {
                    "coreutils": 9.3
                },
                "GNU_SORT_B": {
                    "coreutils": 9.3
                },
                "GNU_SORT_C": {
                    "coreutils": 9.3
                },
                "GRAPHOVERALLCOVERAGE": {
                    "perl": "(v5.26.2))",
                    "graph_overall_coverage.pl": 1.0
                },
                "MINIMAP2_ALIGN": {
                    "minimap2": "2.28-r1209",
                    "samtools": 1.2
                },
                "PRETEXTMAP_HIGHRES": {
                    "pretextmap": "0.1.9",
                    "samtools": 1.21
                },
                "PRETEXTMAP_STANDRD": {
                    "pretextmap": "0.1.9",
                    "samtools": 1.21
                },
                "PRETEXT_INGEST_SNDRD": {
                    "PretextGraph": "0.0.8",
                    "PretextMap": "0.1.9",
                    "bigWigToBedGraph": 447
                },
                "SAMTOOLS_FAIDX": {
                    "samtools": 1.21
                },
                "SAMTOOLS_MERGE": {
                    "samtools": 1.21
                },
                "SAMTOOLS_SORT": {
                    "samtools": 1.21
                },
                "SAMTOOLS_VIEW_FILTER_PRIMARY": {
                    "samtools": 1.21
                },
                "SEQTK_CUTN": {
                    "seqtk": "1.4-r122"
                },
                "SNAPSHOT_SRES": {
                    "pretextsnapshot": "0.0.4"
                },
                "UCSC_BEDGRAPHTOBIGWIG": {
                    "ucsc": 469
                },
                "WINDOWMASKER_MKCOUNTS": {
                    "windowmasker": "1.0.0"
                },
                "WINDOWMASKER_USTAT": {
                    "windowmasker": "1.0.0"
                },
                "Workflow": {
                    "nf-core/curationpretext": "v1.2.0"
                }
            },
            [
                "pipeline_info",
                "pipeline_info/sanger-tol_curationpretext_software_versions.yml",
                "pretext_maps_processed",
                "pretext_maps_processed/CurationPretextTest_hr.pretext",
                "pretext_maps_processed/CurationPretextTest_normal.pretext"
            ],
            5,
            [
                
            ],
            0,
            2,
            true,
            true,
            false
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-03-27T15:09:51.371631753"
    }
}